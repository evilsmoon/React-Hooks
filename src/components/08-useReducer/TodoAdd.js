import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handelAddTodo }) => {

    const [{ description }, handelInputChange, reset] = useForm({
        description: ""
    })

    const handelSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newToDo = {
            id: new Date().getTime(),
            desc: description,
            done: false

        }
        handelAddTodo(newToDo);
        reset();

    }
    return (

        <>
            <h4>add ToDo</h4>
            <hr />
            <form onSubmit={handelSubmit}>
                <input
                    type='text'
                    name='description'
                    placeholder='add..'
                    autoComplete='off'
                    className='form-control'
                    value={description}
                    onChange={handelInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-primary mt-1 btn-block'
                >
                    Add
               </button>
            </form>
        </>
    )
}
