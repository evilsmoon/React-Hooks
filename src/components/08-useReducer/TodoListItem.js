import React from 'react'
import './styles.css'

export const TodoListItem = ({ todo, index, handelToggle ,handelDelete}) => {
    // console.log(todo, i)
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'} `}
                onClick={() => {
                    handelToggle(todo.id)

                }}
            >
                {index + 1}. {todo.desc}
            </p>
            <button
                className='btn btn-danger'
                onClick={() => {
                    handelDelete(todo.id)
                }}
            >
                delete
            </button>
        </li>
    )
}
