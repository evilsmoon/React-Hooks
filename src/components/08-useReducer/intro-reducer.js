
const initialState =[{
    id:1,
    todo:'buy weed',
    done:false
}]

const todoReducer = (state = initialState,action)=>{
   
   if(action?.type === 'add'){
       
       return [...state,action.payload];
   }
   
}
let todos = todoReducer();

// todos.push() no usar push en los arreglos por que muta

const newTodo = {
    id:2,
    todo:'smook weed',
    done:false
}

const addTodoAction = {
    type:'add',
    payload: newTodo
}
todos = todoReducer( todos,addTodoAction );
console.log(todos)

