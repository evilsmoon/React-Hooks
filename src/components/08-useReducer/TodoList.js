import React from 'react';
import { TodoListItem } from './TodoListItem';
export const TodoList = ({ todos ,handelDelete,handelToggle}) => {

    console.log(todos)
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => (
                    
                    <TodoListItem 
                    key={ todo.id }
                    todo={ todo } 
                    index={ i } 
                    handelToggle={handelToggle} 
                    handelDelete={handelDelete}/>
                    
                ))
            }

        </ul>
    )
}

// TodoList.propTypes = {

// }

