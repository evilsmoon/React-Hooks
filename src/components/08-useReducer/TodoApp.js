import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';

import './styles.css'
import { TodoAdd } from './TodoAdd';

const init = ()=> {
    // return [{
    // id: new Date().getTime(),
    // desc: 'aprendiendo React',
    // done : false
    // }];
    return JSON.parse(localStorage.getItem('todos')) || [];
}
export const TodoApp = () => {
    
    
    const [todos,dispatch] = useReducer(todoReducer, [],init);
    
    

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos]);


    const handelDelete = ( todoId ) =>{
        const action = {
            type: 'delete',
            payload:todoId 

            
        }
        dispatch(action);
   
    }

    const handelToggle = ( todoId ) =>{

        dispatch({
            type:'toggle',
            payload:todoId
        })
    }

    const handelAddTodo = ( newToDo) =>{
        dispatch({
            type: 'add',
            payload: newToDo
        } )
    }
    

    return (
        <div>
            <h1>Todo <small>{todos.length}</small></h1>
            <hr/>
        <div className='row'>
            <div className='col-7'>
            <>
            <TodoList 
                    todos={todos} 
                    handelDelete={handelDelete} 
                    handelToggle={handelToggle}/>
            </>

            </div>
            <div className='col-5'>
                <TodoAdd 
                    handelAddTodo={handelAddTodo}/>
            </div>
        </div>
       
        </div>
    )
}
