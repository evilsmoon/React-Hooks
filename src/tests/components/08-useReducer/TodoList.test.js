import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from './fixtures/demoTodos';
describe('test <TodoList/>', () => {
    
    test('debe de mostrar correctamente <TodoList/>', () => {
        
        const todos = demoTodos 
        const handelDelete = jest.fn()
        const handelToggle = jest.fn()
        const wrapper = shallow( <TodoList
            todos={todos}
            handelDelete={handelDelete}
            handelToggle={handelToggle}
        /> )

        expect( wrapper ).toMatchSnapshot();
        
    })
    test('debe de tener el numero de elementos de mi todos', () => {
        
        const todos = demoTodos 
        const handelDelete = jest.fn()
        const handelToggle = jest.fn()
        const wrapper = shallow( <TodoList
            todos={todos}
            handelDelete={handelDelete}
            handelToggle={handelToggle}
        /> )

        expect( wrapper.find('TodoListItem').length ).toBe(todos.length);
        

        expect(wrapper.find('TodoListItem').at(0).prop('handelDelete')).toEqual(expect.any(Function))
    })
    

})
