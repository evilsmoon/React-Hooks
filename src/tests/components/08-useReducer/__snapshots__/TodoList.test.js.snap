// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <TodoList/> debe de mostrar correctamente <TodoList/> 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handelDelete={[MockFunction]}
    handelToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "aprender react",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handelDelete={[MockFunction]}
    handelToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "aprender mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
